src/lib.rs

```diff
@@ -923,118 +901,53 @@ fn format_option(opt: &OptGroup) -> String {
-fn each_split_within<'a, F>(ss: &'a str, lim: usize, mut it: F)
-                            -> bool where F: FnMut(&'a str) -> bool {
-    // Just for fun, let's write this as a state machine:
-
-    let mut slice_start = 0;
-    let mut last_start = 0;
-    let mut last_end = 0;
-    let mut state = A;
-    let mut fake_i = ss.len();
-    let mut lim = lim;
-
-    let mut cont = true;
-
-    // if the limit is larger than the string, lower it to save cycles
-    if lim >= fake_i {
-        lim = fake_i;
-    }
-
-    let mut machine = |cont: &mut bool, state: &mut SplitWithinState, (i, c): (usize, char)| {
-        let whitespace = if c.is_whitespace() { Ws }       else { Cr };
-        let limit      = if (i - slice_start + 1) <= lim  { UnderLim } else { OverLim };
-
-        *state = match (*state, whitespace, limit) {
-            (A, Ws, _)        => { A }
-            (A, Cr, _)        => { slice_start = i; last_start = i; B }
-
-            (B, Cr, UnderLim) => { B }
-            (B, Cr, OverLim)  if (i - last_start + 1) > lim => {
-                // A single word has gone over the limit.  In this
-                // case we just accept that the word will be too long.
-                B
-            }
-            (B, Cr, OverLim)  => {
-                *cont = it(&ss[slice_start..last_end]);
-                slice_start = last_start;
-                B
-            }
-            (B, Ws, UnderLim) => {
-                last_end = i;
-                C
-            }
-            (B, Ws, OverLim)  => {
-                last_end = i;
-                *cont = it(&ss[slice_start..last_end]);
-                A
-            }
-
-            (C, Cr, UnderLim) => {
-                last_start = i;
-                B
-            }
-            (C, Cr, OverLim)  => {
-                *cont = it(&ss[slice_start..last_end]);
-                slice_start = i;
-                last_start = i;
-                last_end = i;
-                B
+fn each_split_within(desc: &str, lim: usize) -> Vec<String> {
+    let mut rows = Vec::new();
+    for line in desc.trim().lines() {
+        let line_chars = line.chars().chain(Some(' '));
+        let words = line_chars.fold( (Vec::new(), 0, 0), |(mut words, a, z), c | {
+            let idx = z + c.len_utf8(); // Get the current byte offset
+
+            // If the char is whitespace, advance the word start and maybe push a word
+            if c.is_whitespace() {
+                if a != z {
+                    words.push(&line[a..z]);
+                }
+                (words, idx, idx)
             }
-            (C, Ws, OverLim)  => {
-                *cont = it(&ss[slice_start..last_end]);
-                A
+            // If the char is not whitespace, continue, retaining the current
+            else {
+                (words, a, idx)
             }
-            (C, Ws, UnderLim) => {
-                C
+        }).0;
+
+        let mut row = String::new();
```

Comment:
We could cut down some more allocations in this part of the function too. We don't need the `filter` anymore because the `words` we get from above are all greater than `0` in length. So we could do something like this:

```rust
let mut current_row = String::new();
for word in words.iter() {
    let sep = if current_row.len() > 0 { Some(" ") } else { None };

    let mut width =
        current_row.width() + word.width() + sep.map(UnicodeWidthStr::width).unwrap_or(0);

    if width <= lim {
        if let Some(sep) = sep {
            current_row.push_str(sep);
        }
        current_row.push_str(word);

        continue
    }

    if current_row.len() > 0 {
        rows.push(current_row.clone());
        current_row.clear();
    }

    current_row.push_str(word);
}

if current_row.len() > 0 {
    rows.push(current_row);
}
```

So we re-use the same `current_row` with its capacity already set somewhere up around `lim` instead of creating a new string buffer each time.

We also don't need to filter and copy `rows`, we can just return it as-is at the end of the method because it's only got valid rows in it.

What do you think?

Reply:
Thanks for the fix. 👍 amended the PR.